{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fd09f948-9270-4b04-820a-58e13e895669",
   "metadata": {},
   "source": [
    "# EXTRACTION OF BREAST CANCER ASSOCIATED TARGETS BIOLOGICAL DATA\n",
    "\n",
    "This Jupyter notebook serves as a comprehensive tool for analyzing gene expression data, particularly focusing on breast cancer research. It is structured to efficiently query, process, and visualize complex datasets, making it a valuable resource for researchers and data analysts in the field of genomics and bioinformatics.\n",
    "\n",
    "Key Features and Functions:\n",
    "\n",
    "**BigQuery Integration for Data Retrieval:**\n",
    "The notebook includes SQL queries to fetch gene expression data from the BigQuery public data repository. It specifically targets RNA expression data relevant to breast cancer studies, filtering and extracting essential information like gene IDs, RNA values, log2 fold changes, study IDs, and contrasts.\n",
    "\n",
    "**Data Processing and Transformation:**\n",
    "A series of custom Python functions are utilized to process the fetched data. This includes handling and cleaning subcellular location data and transforming differential expression data into a more analyzable format. Special attention is given to handling various data scenarios, such as missing or incomplete records.\n",
    "\n",
    "**Differential Expression Analysis:**\n",
    "The notebook performs detailed differential expression analysis. It involves extracting fold changes and contrasts from the data, followed by a logarithmic transformation of RNA values to normalize the data for further analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc7385e5-cf9e-4198-a10d-b22978314a31",
   "metadata": {},
   "source": [
    "# 1 - Import Libraries "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b4cfb89e-9037-47f4-a94b-2e4e9bdbbbb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from google.cloud import bigquery\n",
    "from google.oauth2 import service_account\n",
    "import os\n",
    "import pandas as pd\n",
    "import ast\n",
    "import numpy as np\n",
    "import plotly.express as px\n",
    "\n",
    "from langchain import FewShotPromptTemplate\n",
    "\n",
    "from langchain_experimental.agents.agent_toolkits import create_pandas_dataframe_agent\n",
    "from langchain.chat_models import ChatOpenAI\n",
    "from langchain.agents.agent_types import AgentType\n",
    "\n",
    "from langchain.chat_models import ChatOpenAI\n",
    "\n",
    "from langchain import LLMChain\n",
    "from langchain.chains import SequentialChain\n",
    "\n",
    "# templates\n",
    "from langchain.prompts import (\n",
    "    ChatPromptTemplate,\n",
    "    PromptTemplate,\n",
    "    SystemMessagePromptTemplate,\n",
    "    AIMessagePromptTemplate,\n",
    "    HumanMessagePromptTemplate,\n",
    ")\n",
    "\n",
    "# Schemas\n",
    "\n",
    "from langchain.output_parsers import StructuredOutputParser, ResponseSchema\n",
    "\n",
    "# Progress bar\n",
    "from tqdm.auto import tqdm\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5489e681-58c1-43ff-bfdb-9be2eefbb5d6",
   "metadata": {},
   "source": [
    "# 2 - Functions\n",
    "\n",
    "In this section, we define a series of Python functions that are crucial for our application's operation. \n",
    "Each function is designed to perform a specific task within our overall workflow.\n",
    "\n",
    "**get_service_account_key_path:**\n",
    "Purpose: Retrieves the path to the service account key file. This is likely used for authentication purposes when connecting to external services like Google Cloud services or BigQuery.\n",
    "\n",
    "**create_credentials:**\n",
    "Purpose: Establishes credentials for accessing Google Cloud services. This function likely uses the service account key file to authenticate and generate the necessary credentials.\n",
    "\n",
    "**create_client:**\n",
    "Purpose: Creates a client for interfacing with a specific Google Cloud service, possibly BigQuery. This client is used to execute queries and access data within the cloud environment.\n",
    "\n",
    "**generate_sql_query:**\n",
    "Purpose: Utilizes a Large Language Model to generate SQL queries from natural language input. This function is key in transforming user questions or requests into executable SQL queries.\n",
    "\n",
    "**run_query:**\n",
    "Purpose: Executes an SQL query using the BigQuery client. This function might handle the sending of the query, the reception of results, and any error handling related to the query execution.\n",
    "\n",
    "**dataframe_to_natural_language:**\n",
    "Purpose: Converts data in a pandas DataFrame into a natural language format. This function could be used for summarizing or describing data in a human-readable form.\n",
    "\n",
    "**select_disease_ids_from_dataframe:**\n",
    "Purpose: Extracts disease IDs from a DataFrame. This function is likely specific to the biological data being analyzed and is used to filter or categorize data based on disease identifiers.\n",
    "\n",
    "**query_linked_targets_and_expression:**\n",
    "Purpose: Performs a query to retrieve linked targets and their expression levels from a database, presumably related to biological or genetic data.\n",
    "\n",
    "**main:**\n",
    "Purpose: The main function of the program, orchestrating the overall workflow. It likely calls the other functions in a logical sequence, handling inputs and directing the flow of data processing and analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f905e457-0556-4447-8833-91f2c47895e2",
   "metadata": {},
   "source": [
    "# 3 - Declaring inputs \n",
    "\n",
    "This part of the notebook is where we set up the inputs that will feed into our Large Language Model and BigQuery. Properly defining these inputs is critical for the accuracy and relevance of our results.\n",
    "\n",
    "Inputs Include:\n",
    "\n",
    "Parameters for the LLM, such as prompt details or tuning settings.\n",
    "\n",
    "Credentials and configuration details for BigQuery access.\n",
    "\n",
    "Specific details about the biological data set to be queried, including any filters or criteria for data selection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2bc4b9e8-c1e6-42f3-a795-b56d0efe1680",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The results for your query are:\n",
      "             id                                                         name\n",
      "  MONDO_0007254                                                breast cancer\n",
      "Orphanet_227535                                     Hereditary breast cancer\n",
      "    EFO_0005603                               age at breast cancer diagnosis\n",
      "    EFO_0008040 breast cancer anti-estrogen resistance protein 3 measurement\n",
      "    EFO_0009782                 progesterone-receptor positive breast cancer\n",
      "    EFO_0005537                                triple-negative breast cancer\n",
      "    EFO_0009443                                          BRCAX breast cancer\n",
      "    EFO_0009649                              susceptibility to breast cancer\n",
      "    EFO_1000650                     estrogen-receptor negative breast cancer\n",
      "    EFO_0009781                 progesterone-receptor negative breast cancer\n",
      "    EFO_1000649                     estrogen-receptor positive breast cancer\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter the disease IDs separated by commas (e.g., ID1,ID2):  MONDO_0007254, EFO_0005603, EFO_0009782, EFO_0009781\n"
     ]
    }
   ],
   "source": [
    "# Constants\n",
    "SERVICE_ACCOUNT_FOLDER_NAME = 'your current working directory'\n",
    "SERVICE_ACCOUNT_FILE_NAME = 'your big query service account'\n",
    "os.environ['OPENAI_API_KEY'] = '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'\n",
    "\n",
    "# Define response schemas\n",
    "response_schemas = [\n",
    "        ResponseSchema(name=\"sql\", \n",
    "                       description=\"Generate a SQL query to output a disease_id and a disease_name from open targets.\"),\n",
    "]\n",
    "\n",
    "# System and human message templates\n",
    "system_message = '''\n",
    "\n",
    "Task: \n",
    "Generate SQL queries as JSON objects based on the provided instructions\n",
    "\n",
    "Instructions:\n",
    "\n",
    "- bigquery-public-data.open_targets_platform.diseases\n",
    "- Use the following fields:  id, name\n",
    "- In the SELECT clause of the SQL query, include \"id\" and \"name\"\n",
    "- Do not use any other table names, column names, or data types that are not provided\n",
    "- If you cannot generate an SQL query based on the provided schema, explain the reason to the user\n",
    "\n",
    "\n",
    "Example: \"What is the gene expression profile associated with <<disease name>>?\"\n",
    "\n",
    "  Correct Usage:\n",
    "  \n",
    "    SELECT id, name\n",
    "    FROM `bigquery-public-data.open_targets_platform.diseases`\n",
    "    WHERE LOWER(name) LIKE '%<<disease name>>%';\n",
    "\n",
    "'''\n",
    "human_message = human_message = 'Answer the users question as best as possible.\\n{format_instructions}\\n{question}'\n",
    "\n",
    "# Question to be asked\n",
    "question = 'What are targets associated with breast cancer?'\n",
    "\n",
    "# Call the main function with all necessary arguments and get the natural language output\n",
    "output = main(question, response_schemas, system_message, human_message)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b1ca1ffc-48aa-4fd0-a4ed-1cdc552dc8bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>tissue_label</th>\n",
       "      <th>rna_value</th>\n",
       "      <th>protein_level</th>\n",
       "      <th>subcellularLocations</th>\n",
       "      <th>isInMembrane</th>\n",
       "      <th>isSecreted</th>\n",
       "      <th>tissueSpecificity</th>\n",
       "      <th>tissueDistribution</th>\n",
       "      <th>diseaseIds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000175471</td>\n",
       "      <td>breast</td>\n",
       "      <td>1500.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>{'list': [{'element': {'location': 'Cytoplasmi...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[MONDO_0007254]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000104321</td>\n",
       "      <td>breast</td>\n",
       "      <td>5.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>{'list': [{'element': {'location': 'Cell membr...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>[MONDO_0007254]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000221716</td>\n",
       "      <td>breast</td>\n",
       "      <td>8.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>None</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[MONDO_0007254]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000140396</td>\n",
       "      <td>breast</td>\n",
       "      <td>2680.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>{'list': [{'element': {'location': 'Nucleus', ...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>[MONDO_0007254]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000164778</td>\n",
       "      <td>breast</td>\n",
       "      <td>9.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>{'list': [{'element': {'location': 'Nucleus', ...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[MONDO_0007254]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7586</th>\n",
       "      <td>ENSG00000132446</td>\n",
       "      <td>breast</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[MONDO_0007254]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7587</th>\n",
       "      <td>ENSG00000164399</td>\n",
       "      <td>breast</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>{'list': [{'element': {'location': 'Secreted',...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[MONDO_0007254]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7588</th>\n",
       "      <td>ENSG00000236481</td>\n",
       "      <td>breast</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>None</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[MONDO_0007254]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7589</th>\n",
       "      <td>ENSG00000215262</td>\n",
       "      <td>breast</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>{'list': [{'element': {'location': 'Cell membr...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[MONDO_0007254]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7590</th>\n",
       "      <td>ENSG00000167346</td>\n",
       "      <td>breast</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>{'list': [{'element': {'location': 'Secreted',...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[MONDO_0007254]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7591 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   id tissue_label  rna_value  protein_level  \\\n",
       "0     ENSG00000175471       breast     1500.0             -1   \n",
       "1     ENSG00000104321       breast        5.0             -1   \n",
       "2     ENSG00000221716       breast        8.0             -1   \n",
       "3     ENSG00000140396       breast     2680.0             -1   \n",
       "4     ENSG00000164778       breast        9.0             -1   \n",
       "...               ...          ...        ...            ...   \n",
       "7586  ENSG00000132446       breast        0.0             -1   \n",
       "7587  ENSG00000164399       breast        0.0             -1   \n",
       "7588  ENSG00000236481       breast        0.0             -1   \n",
       "7589  ENSG00000215262       breast        1.0             -1   \n",
       "7590  ENSG00000167346       breast        0.0             -1   \n",
       "\n",
       "                                   subcellularLocations  isInMembrane  \\\n",
       "0     {'list': [{'element': {'location': 'Cytoplasmi...             0   \n",
       "1     {'list': [{'element': {'location': 'Cell membr...             1   \n",
       "2                                                  None          <NA>   \n",
       "3     {'list': [{'element': {'location': 'Nucleus', ...             0   \n",
       "4     {'list': [{'element': {'location': 'Nucleus', ...             0   \n",
       "...                                                 ...           ...   \n",
       "7586                                       {'list': []}          <NA>   \n",
       "7587  {'list': [{'element': {'location': 'Secreted',...             0   \n",
       "7588                                               None          <NA>   \n",
       "7589  {'list': [{'element': {'location': 'Cell membr...             1   \n",
       "7590  {'list': [{'element': {'location': 'Secreted',...             0   \n",
       "\n",
       "      isSecreted  tissueSpecificity  tissueDistribution       diseaseIds  \n",
       "0              0                0.5                 0.0  [MONDO_0007254]  \n",
       "1              0                0.5                 0.5  [MONDO_0007254]  \n",
       "2           <NA>                NaN                 NaN  [MONDO_0007254]  \n",
       "3              0               -1.0                -1.0  [MONDO_0007254]  \n",
       "4              0                1.0                 1.0  [MONDO_0007254]  \n",
       "...          ...                ...                 ...              ...  \n",
       "7586        <NA>                1.0                 1.0  [MONDO_0007254]  \n",
       "7587           1                NaN                 NaN  [MONDO_0007254]  \n",
       "7588        <NA>                NaN                 NaN  [MONDO_0007254]  \n",
       "7589           0                1.0                 1.0  [MONDO_0007254]  \n",
       "7590           1                1.0                 1.0  [MONDO_0007254]  \n",
       "\n",
       "[7591 rows x 10 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output.drop(columns=['targetId'], inplace=True)\n",
    "output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "554a309e-630f-453b-ac77-8f8e22325497",
   "metadata": {},
   "source": [
    "# PART 2 - Query differential expression data\n",
    "\n",
    "This notebook includes a script designed to query and retrieve specialized gene expression data from a BigQuery database, focusing specifically on studies related to breast cancer. The script executes an SQL query against the open_targets_platform dataset within the BigQuery public data repository. This query selectively retrieves a range of fields including identifiers like datasourceId, datatypeId, targetId, and information about the studies such as contrast, studyOverview, and expression values like log2FoldChangeValue.\n",
    "\n",
    "The query is specifically tailored to extract data where the datatypeId is 'rna_expression', ensuring that the results are relevant to RNA-based studies. Additionally, the records are filtered to include those that either mention 'breast', 'bc' (likely shorthand for breast cancer), or reference 'breast cancer' in their study overview. This targeted approach allows us to focus on gene expression data that is most relevant to breast cancer research.\n",
    "\n",
    "To enable secure and authenticated access to the BigQuery service, the script first retrieves the necessary service account key path, generates authentication credentials, and then establishes a connection to the BigQuery client. The results of the query are then fetched and stored for further analysis or visualization within the notebook, providing valuable insights into RNA expression patterns in breast cancer studies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "540764d4-818a-4795-ae34-1ff1c70e22d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             targetId                                        diffexpdata\n",
      "0     ENSG00000000003  [{'contrast': ''breast cancer' vs 'normal'', '...\n",
      "1     ENSG00000000005  [{'contrast': ''invasive ductal carcinoma' vs ...\n",
      "2     ENSG00000000419  [{'contrast': ''invasive ductal carcinoma' vs ...\n",
      "3     ENSG00000000938  [{'contrast': ''invasive ductal carcinoma' vs ...\n",
      "4     ENSG00000000971  [{'contrast': ''breast carcinoma' vs 'normal''...\n",
      "...               ...                                                ...\n",
      "5531  ENSG00000286125  [{'contrast': ''breast cancer' vs 'normal'', '...\n",
      "5532  ENSG00000286214  [{'contrast': ''breast carcinoma' vs 'normal''...\n",
      "5533  ENSG00000286522  [{'contrast': ''breast cancer; breast' vs 'nor...\n",
      "5534  ENSG00000287080  [{'contrast': ''breast cancer; breast' vs 'nor...\n",
      "5535  ENSG00000288569  [{'contrast': ''breast cancer; breast' vs 'nor...\n",
      "\n",
      "[5536 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# Initialize an empty list to store the new formatted data\n",
    "# Group by 'targetId' and aggregate the data\n",
    "grouped_exp = exp_data.groupby('targetId').apply(lambda x: x[['contrast', 'log2FoldChangeValue', 'studyId', 'StudyOverview']].to_dict('records')).reset_index()\n",
    "\n",
    "# Rename the columns for clarity\n",
    "grouped_exp.columns = ['targetId', 'diffexpdata']\n",
    "\n",
    "# Display the new DataFrame\n",
    "print(grouped_exp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2f829a1-72db-4d2c-84bd-5bd08b2ce809",
   "metadata": {},
   "source": [
    "# PART 3 - Merging outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4d67ee5e-ba6e-4733-a10f-4215271c07ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df = pd.merge(output, grouped_exp, left_on='id', right_on='targetId', how = 'left')\n",
    "final_df.drop(columns = ['targetId','tissue_label'],inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ed658d17-149c-44c0-af7b-1c81ae315875",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df.rename(columns = {'rna_value':'rna_value_breast'}, inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acf81ba5-d24c-4754-b422-1191daf4b7c0",
   "metadata": {},
   "source": [
    "# PART 4 - Cleaning data\n",
    "\n",
    "## 4.1 - subcellular location\n",
    "\n",
    "In this section of the notebook, we process the subcellularLocations column of the final_df DataFrame to extract and clean up location data. The code iterates through each record in the subcellularLocations column, handling various scenarios to ensure robust data processing. The primary goal is to compile a list (lst) of cleaned location data, which is then appended back to final_df as a new column locations_clean.\n",
    "\n",
    "The processing covers four distinct cases:\n",
    "\n",
    "**No Data in Record**: If the record itself is None, indicating an absence of data, we append the phrase \"No location data\" to our list.\n",
    "\n",
    "**Missing or None 'list' Key**: If the key 'list' is either not present in the record or its value is None, we again append \"No location data\".\n",
    "\n",
    "**Empty 'list'**: If the 'list' key exists but contains an empty list, \"No location data\" is appended, signifying a lack of specific location information.\n",
    "\n",
    "**Valid Data in 'list'**: In cases where 'list' contains elements, we extract the 'location' field from each element and join these locations into a single string separated by commas. This string is then appended to lst.\n",
    "\n",
    "After processing all records, we check the length of lst to ensure it aligns with our expectations and then assign this list as a new column (locations_clean) in final_df. This approach effectively cleans and structures the subcellular location data, making it more accessible and meaningful for further analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "716f2685-161a-4613-82ce-610877d68b1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7591\n"
     ]
    }
   ],
   "source": [
    "lst = []\n",
    "\n",
    "for record in final_df['subcellularLocations']:\n",
    "    # Case 1: Record itself is None\n",
    "    if record is None:\n",
    "        lst.append(\"No location data\")\n",
    "        continue\n",
    "\n",
    "    # Case 2: 'list' key does not exist in the record or 'list' is None\n",
    "    if 'list' not in record or record['list'] is None:\n",
    "        lst.append(\"No location data\")\n",
    "        continue\n",
    "\n",
    "    # Case 3: 'list' exists but is empty\n",
    "    if len(record['list']) == 0:\n",
    "        lst.append(\"No location data\")\n",
    "        continue\n",
    "\n",
    "    # Case 4: Normal case where 'list' contains elements\n",
    "    temp_lst = [dct['element']['location'] for dct in record['list']]\n",
    "    lst.append(','.join(temp_lst))\n",
    "\n",
    "# Check the length of lst\n",
    "print(len(lst))\n",
    "\n",
    "final_df['locations_clean'] = lst"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fb9296e-0b42-4cb3-920b-e9c8e162b09f",
   "metadata": {},
   "source": [
    "## 4.2 - Formating differential expression data\n",
    "\n",
    "In this part of the notebook, we're dealing with the extraction and processing of gene expression data. The core of this section is a function named extract_fold_changes_and_contrasts. This function is designed to work with each row in a DataFrame, focusing specifically on the data in the 'diffexpdata' column.\n",
    "\n",
    "The function first checks if 'diffexpdata' in each row is a list. If it is, the function processes each item in this list, which are expected to be dictionaries containing key data points like gene 'id', RNA expression value 'rna_value_breast', study ID 'studyId', the log2 fold change value 'log2FoldChangeValue', and the experimental 'contrast'. It repackages this data into a new dictionary format. If 'diffexpdata' isn't a list (likely meaning it's NaN or missing), the function handles this by creating a dictionary with default values indicating missing data.\n",
    "\n",
    "After applying this function to each row of our main DataFrame, we concatenate the results into a new DataFrame called 'long_df'. This DataFrame is structured in a long format, which is generally more useful for data analysis, especially in complex datasets like gene expression studies.\n",
    "\n",
    "Additionally, we perform a logarithmic transformation on the RNA expression values to normalize this data, helping to reveal more intricate patterns and relationships. This transformed data is stored in a new column, 'Log_RNA_Value_Breast', in 'long_df'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "246a5762-b967-49c4-b0cd-7afb8c1586f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_fold_changes_and_contrasts(row):\n",
    "    # Check if diffexpdata is a list\n",
    "    if isinstance(row['diffexpdata'], list):\n",
    "        # Process the list of dictionaries in diffexpdata\n",
    "        return [{'id': row['id'], 'rna_value_breast': row['rna_value_breast'],'study_id' : item.get('studyId', None),\n",
    "                 'Log2FoldChangeValue': item.get('log2FoldChangeValue', float('nan')), 'Contrast': item.get('contrast', None)}\n",
    "                for item in row['diffexpdata']]\n",
    "    else:\n",
    "        # Handle the case where diffexpdata is not a list (presumably NaN)\n",
    "        return [{'id': row['id'], 'rna_value_breast': row['rna_value_breast'], 'study_id' : None,\n",
    "                 'Log2FoldChangeValue': float('nan'), 'Contrast': None}]\n",
    "\n",
    "# Applying the function and creating a long-format DataFrame\n",
    "long_df = pd.concat([pd.DataFrame(extract_fold_changes_and_contrasts(row)) for _, row in final_df.iterrows()], ignore_index=True)\n",
    "# Apply logarithmic transformation to RNA_Value_Breast\n",
    "long_df['Log_RNA_Value_Breast'] = np.log(long_df['rna_value_breast'] + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "06f82ecd-8915-42b9-84d4-5575985a174b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>rna_value_breast</th>\n",
       "      <th>study_id</th>\n",
       "      <th>Log2FoldChangeValue</th>\n",
       "      <th>Contrast</th>\n",
       "      <th>Log_RNA_Value_Breast</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000175471</td>\n",
       "      <td>1500.0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>7.313887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000104321</td>\n",
       "      <td>5.0</td>\n",
       "      <td>E-GEOD-52194</td>\n",
       "      <td>5.4</td>\n",
       "      <td>'HER2 Positive Breast Carcinoma; breast carcin...</td>\n",
       "      <td>1.791759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000221716</td>\n",
       "      <td>8.0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>2.197225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000140396</td>\n",
       "      <td>2680.0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>7.893945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000164778</td>\n",
       "      <td>9.0</td>\n",
       "      <td>E-MTAB-779</td>\n",
       "      <td>1.5</td>\n",
       "      <td>'breast carcinoma' vs 'normal'</td>\n",
       "      <td>2.302585</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                id  rna_value_breast      study_id  Log2FoldChangeValue  \\\n",
       "0  ENSG00000175471            1500.0          None                  NaN   \n",
       "1  ENSG00000104321               5.0  E-GEOD-52194                  5.4   \n",
       "2  ENSG00000221716               8.0          None                  NaN   \n",
       "3  ENSG00000140396            2680.0          None                  NaN   \n",
       "4  ENSG00000164778               9.0    E-MTAB-779                  1.5   \n",
       "\n",
       "                                            Contrast  Log_RNA_Value_Breast  \n",
       "0                                               None              7.313887  \n",
       "1  'HER2 Positive Breast Carcinoma; breast carcin...              1.791759  \n",
       "2                                               None              2.197225  \n",
       "3                                               None              7.893945  \n",
       "4                     'breast carcinoma' vs 'normal'              2.302585  "
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "long_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4ea0c42-1ef5-442a-bf6a-23e582b5da0d",
   "metadata": {},
   "source": [
    "# 5 - Saving outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "42cc6d9c-858e-41ae-abd1-df9474149cae",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df[['id','diseaseIds','isInMembrane','locations_clean']].to_csv('TargetsLocations.csv',index=False)\n",
    "long_df.to_csv('diffexpdata.csv',index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "google_batch",
   "language": "python",
   "name": "google_batch"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
